python/ray/tune/pbt.py

```diff
@@ -0,0 +1,174 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import numpy as np
+import random
+import math
+import copy
+
+from ray.tune.trial_scheduler import FIFOScheduler, TrialScheduler
+from ray.tune.trial import Trial
+import ray
+class PopulationBasedTraining(FIFOScheduler):
+    """Implements the Population Based Training algorithm as described in the PBT paper:
+
+    Args:
+        time_attr (str): The TrainingResult attr to use for documenting length
+            of time since last ready() call. Attribute only has to increase
+            monotonically
+        reward_attr (str): The TrainingResult objective value attribute. As with
+            'time_attr'. this may refer to any objective value that is supposed
+            to increase with time.
+        grace_period (str): Period of time, in which algorithm will not compare 
+            model to other models
+        ready (func): Determines in Trial is ready to undergo the comparison,
+            exploitation, and exploration process. Should return boolean
+        hyperparamter_mutations (dict); Possible values that each hyperparameter can mutate
+            to, as certain hyperparameters only work with certain values
+    """
+    
+    def __init__(
+            self, time_attr="training_iteration", reward_attr='episode_reward_mean',
+            grace_period=4.0, ready=None, hyperparameter_mutations=None):
+        FIFOScheduler.__init__(self)
+        self._stopped_trials = set()
+        self._completed_trials = set()
+        self._results = collections.defaultdict(list)
+        self._time_tracker = {}
+        self._grace_period = grace_period
+        self._reward_attr = reward_attr
+        self._time_attr = time_attr
+        self._ready_func = ready
+        self._hyperparameter_mutations = hyperparameter_mutations
+
+    def on_trial_result(self, trial_runner, trial, result):
+
+        print("wowee, trial had a result")
+        print(getattr(trial, "config"))
+        self._results[trial].append(result)
+        time = getattr(result, self._time_attr)
+        ready = False
+        #check model is ready to undergo mutation, based on user function or default
+        #function
+        if time > self._grace_period:
+            ready = self._ready_func(result, trial) if self._ready_func is not None else self._default_ready(result, trial, time)
+        else:
+            print("grace period")
+            ready = False
+        if ready:
+            print("ready to undergo mutation")
+            print ("----")
+            print("Current Trial is: {0}".format(trial))
+            #get best trial for current time
+            best_trial = self._get_best_trial(result, time)
```

Comment:
Let's make exploit / explore functions as in the paper, and implement the truncation selection rule (pick best trial randomly from top 20%, and only exploit if in bottom 20%). It seems like the truncation selection rule had to best results.